lxd/device/nic_p2p.go

```diff
@@ -0,0 +1,138 @@
+package device
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/lxc/lxd/lxd/device/config"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/shared"
+)
+
+type nicP2P struct {
+	deviceCommon
+}
+
+// valite checks the supplied config for correctness.
+func (d *nicP2P) validate() error {
+	if d.instance.Type() != instance.TypeContainer {
+		return ErrUnsupportedDevType
+	}
+
+	if d.config["ipv4.routes"] != "" {
```

Comment:
This duplication with the bridged nictype makes me wonder if there isn't something we can do to abstract this a bit, at least for nic devices.

Could we make it so we have a shared function for all nic device types which takes in the nictype (in case we need to special case, for example for ipv4.address/ipv6.address), the config, a list of allowed config keys and a list of required config keys.

That one function would then immediately fail if any of the required keys are missing or if any extra key is set, then go on to validate the rest key by key with just the needed special casing here and there.

type & nictype would be silently validate, no need to specifically allow/require them.

That would turn this function into pretty much:

```
func (d *nicP2P) validate() error {
    allowedKeys := []string{"name", "mtu", "hwaddr", "host_name", "limits.ingress", "limits.egress", "limits.max", "ipv4.routes", "ipv6.routes"}
    requiredKeys := []string{}

    return nicValidate(d.config, "p2p", allowedKeys, requiredKeys)
}
```

What do you think of that ^ ?

Reply:
@stgraber overall sounds good, but I would be hesitant of pushing device type logic into a central validation function, after all this PR is trying to reduce the device specific logic being implemented in shared spaces (there are currently only 2 non-device-specific references to nictype, and both are related to p2p/bridged routing).

However I see your point, and I think we could re-use the approach used in https://github.com/lxc/lxd/blob/master/lxd/networks_config.go#L11 by having a map of callback functions that then validate each field by doing field specific logic. I could then add a function called `deviceConfigValidate(rules, d.config)` which could then be used by all devices, not just NICs. I could make this function ignore "type" and "nictype" though to avoid duplication.

I could then define a function that returns a shared set of validation rules for use with p2p and bridge, then each can add any extra fields they need.

Reply:
I've added this sort of validation now.